pkg/apis/metal3/v1alpha1/baremetalhost_types.go

```diff
@@ -853,6 +857,50 @@ func (host *BareMetalHost) GetImageChecksum() (string, string, bool) {
 	}
 }
 
+// ValidateForAdmission returns an error if the host fails uniqueness validation,
+// and cannot be added to the cluster.  This is ordinarily run by an
+// admission webhook.
+func (host *BareMetalHost) ValidateForAdmission(c client.Client) error {
+	ctx := context.Background()
+	hostList := BareMetalHostList{}
+	opts := &client.ListOptions{
+		Namespace: host.Namespace,
+	}
+
+	err := c.List(ctx, &hostList, opts)
+
+	if err != nil {
+		return err
+	}
+
+	var allErrs field.ErrorList
+
+	for _, existingHost := range hostList.Items {
+		if host.Spec.BMC.Address == existingHost.Spec.BMC.Address {
```

Comment:
Do we want to check for `.Address == ""` earlier? For unmanaged hosts, the BMC address will be optional. If this lands first perhaps we defer adding that check until the unmanaged PR lands https://github.com/metal3-io/baremetal-operator/pull/569

Reply:
I'm not sure how think about uniqueness in this context given that multiple fields are nullable. Do we have a list of hard requirements/rules?

Reply:
We want the BMC.Address to be unique, if it is set, to avoid having 2 resources managing the same physical host.

We want the BMC.CredentialsName to be unique, if it is set, to avoid having 2 host resources taking ownership of the same Secret resource.

We want the BootMACAddress to be unique to avoid having 2 resources trying to claim the same data for a discovered node. The current draft of the code says the field can be empty, but I don't think we want to support that given that we've seen bugs recently.

Reply:
I have added a check for the credentials name field.